notes.txt

```diff
@@ -0,0 +1,241 @@
+******************************REWARDS******************************
+*******************************************************************
+
+# SECTION 2
+# 0 Set withdrawal address of delegation account to withdrawal account in host zone
+    registration ack
+# 1 Update DepositRecord to handle rewards types
+# 2 Update epoch end to process all records for each host zone
+# 3 Add new records for rewards that accrued in the previous epoch
+
+>>> EPOCH START
+----------
+... rewards accrue via calls to liquid stake
+
+>>> EPOCH END
+
+# process all unprocessed records for each host zone
+for record in records where record.type == REWARD:
+  def bankSendAck(ack):
+    record = getRecord(ack.id)
+    record.type = DEPOSIT
+    record.status = STAKE
+  # do upon completion
+  ICABankSend.Ack = bankSendAck
+  ICABankSend(from=withdrawal, to=delegat, amount=record.amount)
+
+# add new records for rewards that accrued in the previous epoch
+for hz in getAllHostZones():
+  def addRewardsRecord(hz):
+    # define a query callback function
+    def queryCB(hz):
+      recordBalance = sumAllRewardsRecords()
+      newBalance = balance - recordBalance
+      createNewRewardsRecord(newBalance)
+    wA = hz.withdrawalAccount()
+    query(address=wA, queryType=balance, cb=queryCB)
+
+
+# NOTE: this is executed after processing records because although ICA calls
+# _are_ ordered (meaning we could first send balances from withdrawal -> delegate account),
+# there's no way to enforce that the records are updated on stride without chaining these.
+# unclear whether chaining is a robust approach. What happens if one step fails?
+# what do we have to roll back?
+# easier to think about these in isolation IMO
+1. ICQ wA balance
+2. ICQ CB: create records, BankSend ICA call them from withdrawal -> delegate account
+3. Bank send ICA Ack: update deposit records, iterate through them and MsgDelegate ICA call them
+4. MsgDelegate ICA ack: delete deposit records
+    
+
+----------
+NEXT EPOCH START
+...
+
+
+
+
+
+
+
+
+
+*******************************************************************
+******************************UNBONDING****************************
+*******************************************************************
+
+# SIMPLIFYING ASSUMPTION: Only allow users to process 1 unbonding / block
+# this allows us to more easily track unique unbonding records for a user
+# by keying them as senderAddress:hostZoneId:amount:block
+
+
+# SECTION 3
+# 0 Add UnbondingRecord type
+# 1 Add RedeemStake logic outlined below
+# 2 Update Epoch end to
+  # a. PROCESS RECORDS - ENQUEUE
+  # b. CACHE MOST RECENT BLOCK HEIGHT FOR HOST ZONE
+  # c. PROCESS RECORDS - SEND BACK TO USERS
+
+
+>>> EPOCH START
+----------
+
+# users call redeem stake
+def redeemStake(tokens):
+  unbondingRecord = newUnbondingRecord()
+  # getUnbondingBlockHeight queries the unbonding record
+  # in the creation ack and adds a dequeue unbonding height
```

Comment:
not sure we can do this by block height, might need to use clock time. ideally we can do it by height